### DISCUSSION TITLE
Skip fails Kotlin, unable to build

### DISCUSSION THREAD

**Discussion by anjumkaiser** [skiptools/skip#141] [General]
Hi

I installed skip and ran skip checkup

it fails on Kotlin tests.

skip checkup
[✓] Skip version 0.8.48 (= 0.8.48)
[✓] macOS version 14.5 (> 13.5.0)
[✓] Swift version 5.10 (> 5.9.0)
[✓] Xcode version 15.4 (> 15.0.0)
[✓] Xcode tools SDKs: 7
[✓] Homebrew version 4.3.3 (> 4.1.0)
[✓] Gradle version 8.8 (> 8.6.0)
[✓] Java version 22.0.1 (> 21.0.0)
[✓] Android Debug Bridge version 1.0.41 (> 1.0.40)
[✓] Android Studio version: 2023.3
[✓] Android tools SDKs: 7
[✓] Resolve dependencies (12.3s)
[✓] Build hello-skip (10.34s)
[✓] Test Swift (5.53s)
[✗] Test Kotlin (2.49s)
[✓] Archive iOS ipa (18.88s)
[✓] Assemble HelloSkip-release.ipa 37 KB
[✓] Verify HelloSkip-release.ipa 37 KB
[✗] Assembling Android apk (9.32s)
[✗] Verify HelloSkip-release.apk: The file “app-release-unsigned.apk” couldn’t be opened because there is no such file.
[✓] Check Swift Package (0.14s)
[✗] Skip 0.8.48 checkup failed with 3 errors
Output logged to: /tmp/skip-checkup-2024-06-09T12:11:51Z.txt
See https://skip.tools/docs/faq and https://community.skip.tools for help

I ran skip checkup --verbose its log file is here:

[skip-checkup-2024-06-09T11:50:54Z.txt](https://github.com/user-attachments/files/15751958/skip-checkup-2024-06-09T11.50.54Z.txt)

I did skip init as in document, it created  project but building it in Xcode fails:
[Build HelloWorld_2024-06-09T17-22-59.txt](https://github.com/user-attachments/files/15753998/Build.HelloWorld_2024-06-09T17-22-59.txt)

**Comment by marcprux** [comment]
There appear to be two separate issues here. First the checkup failure:

> FAILURE: Build failed with an exception.
> * What went wrong:
> Could not determine the dependencies of task ':app:generateReleaseLintVitalReportModel'.
> > Could not create task ':app:compileReleaseJavaWithJavac'.
>    > Failed to calculate the value of task ':app:compileReleaseJavaWithJavac' property 'javaCompiler'.
>       > Cannot find a Java installation on your machine matching this tasks requirements: {languageVersion=21, vendor=any, implementation=vendor-specific} for MAC_OS on aarch64.
>          > No locally installed toolchains match and toolchain download repositories have not been configured.
> 

Do you have a custom `JAVA_HOME` set in your environment? You can check with `echo $JAVA_HOME` from the Terminal. Also, can you post the output of `which java` and `java -version`?

Another historic cause of this issue is the scenario where the machine migrated from an Intel Mac to an ARM Mac, but the old Intel-specific Homebrew is still installed. You can identify this situation if you have `/usr/local/homebrew/` (Intel) rather then the ARM version at `/opt/homebrew/`. If this applies in your case, the solution is to install Homebrew fresh from https://brew.sh and then re-install Skip.

Then the Xcode error:

> error: An error occurred while performing transpilation: expected colon, near ""
> Error: expected colon, near ""

This sounds like there might be a format error in your `~/.skiptools/skipkey.env` file. Can you make sure it is in the form:

```
SKIPKEY: key
```

**Reply by anjumkaiser** [reply]
there was a new line before key, I removed it and tried again

skip checkup still fails at Kotlin:

➜  Developer skip checkup
[✓] Skip version 0.8.48 (= 0.8.48)
[✓] macOS version 14.5 (> 13.5.0)
[✓] Swift version 5.10 (> 5.9.0)
[✓] Xcode version 15.4 (> 15.0.0)
[✓] Xcode tools SDKs: 7
[✓] Homebrew version 4.3.4 (> 4.1.0)
[✓] Gradle version 8.8 (> 8.6.0)
[✓] Java version 22.0.1 (> 21.0.0)
[✓] Android Debug Bridge version 1.0.41 (> 1.0.40)
[✓] Android Studio version: 2023.3
[✓] Android tools SDKs: 7
[✓] Resolve dependencies (42.56s)
[✓] Build hello-skip (12.53s)
[✓] Test Swift (6.26s)
[✗] Test Kotlin (10.88s)
[✓] Archive iOS ipa (29.7s)
[✓] Assemble HelloSkip-release.ipa 37 KB
[✓] Verify HelloSkip-release.ipa 37 KB
[✗] Assembling Android apk (18.59s)
[✗] Verify HelloSkip-release.apk: The file “app-release-unsigned.apk” couldn’t be opened because there is no such file.
[✓] Check Swift Package (0.16s)
[✗] Skip 0.8.48 checkup failed with 3 errors
Output logged to: /tmp/skip-checkup-2024-06-10T13:37:01Z.txt
See https://skip.tools/docs/faq and https://community.skip.tools for help
Error: 4 errors

[skip-checkup-2024-06-10T13:37:01Z.txt](https://github.com/user-attachments/files/15774058/skip-checkup-2024-06-10T13.37.01Z.txt)

I again tried with:

skip init --open-xcode --appid=com.xyz.HelloSkip hello-skip HelloSkip

  Developer skip init --open-xcode --appid=com.xyz.HelloSkip hello-skip HelloSkip

Initializing Skip application hello-skip
[✓] Resolve dependencies (12.39s)
[✓] Build hello-skip (10.61s)
[✓] Check Swift Package (0.14s)
[✓] Created module HelloSkip in /Users/anjumkaiser/Developer/hello-skip
[✓] Opening Xcode project (0.09s)
[✓] Skip 0.8.48 init succeeded in 23.28s

I again tried building with Xcode, it gives different error this time
[Build HelloWorld_2024-06-09T17-22-59.txt](https://github.com/user-attachments/files/15774121/Build.HelloWorld_2024-06-09T17-22-59.txt)

I removed openjdk@22 and forced link openjdk@21 but that causes skip grade and android tests to fail.
are there specific dependencies I need to install?

I force removed kotlin@2 and openjdk@22 as they were installed by something else, then set following
~ export JAVA_HOME=/opt/homebrew/Cellar/openjdk@21/21.0.3
~ export PATH=$JAVA_HOME/jre/bin:$JAVA_HOME/bin:$PATH

this cause skip checkup to work:

skip checkup
[✓] Skip version 0.8.48 (= 0.8.48)
[✓] macOS version 14.5 (> 13.5.0)
[✓] Swift version 5.10 (> 5.9.0)
[✓] Xcode version 15.4 (> 15.0.0)
[✓] Xcode tools SDKs: 7
[✓] Homebrew version 4.3.4 (> 4.1.0)
[✓] Gradle version 8.8 (> 8.6.0)
[✓] Java version 21.0.3 (> 21.0.0)
[✓] Android Debug Bridge version 1.0.41 (> 1.0.40)
[✓] Android Studio version: 2023.3
[✓] Android tools SDKs: 7
[✓] Resolve dependencies (12.67s)
[✓] Build hello-skip (10.6s)
[✓] Test Swift (5.69s)
[✓] Test Kotlin (35.78s)
[✓] Archive iOS ipa (27.25s)
[✓] Assemble HelloSkip-release.ipa 37 KB
[✓] Verify HelloSkip-release.ipa 37 KB
[✓] Assembling Android apk (105.18s)
[✓] Verify HelloSkip-release.apk 12.7 MB
[✓] Check Swift Package (0.35s)
[✓] Skip 0.8.48 checkup succeeded in 198.05s

I ran again:

skip init --open-xcode --appid=com.xyz.HelloSkip hello-skip HelloSkip

it fails to build with gradle related issues:
[Build HelloSkip_2024-06-10T19-04-41.txt](https://github.com/user-attachments/files/15774540/Build.HelloSkip_2024-06-10T19-04-41.txt)

seems dependencies will be a problem for now as there is already  kotlin 2 and openjdk@22 is default in homebrew

is there a complete list of packages I need to have to get this to work?

**Comment by anjumkaiser** [comment]
I rebooted my Mac, it seems to skip checkup and skip init now works fine, skip test build  debug build got android too.
but Xcode just fail at the same issue.

**Comment by marcprux** [comment]
I think the Xcode issue is just that you need to launch an Android emulator from the Android Studio Device Manager:

```
GRADLE> e: file://:0:0 No connected Android devices or emulators were reported by `adb devices`. To launch the Skip app, start an emulator from the Android Studio Device Manager or use the ~/Library/Android/sdk/emulator/emulator command
…
GRADLE> * What went wrong:
GRADLE> Execution failed for task ':app:installDebug'.
No connected Android devices or emulators were reported by `adb devices`. To launch the Skip app, start an emulator from the Android Studio Device Manager or use the ~/Library/Android/sdk/emulator/emulator command

GRADLE> > com.android.builder.testing.api.DeviceException: No connected devices!
```

Let us know if that helps.

And thanks for letting us know about the OpenJDK@22/Kotlin2 issue. There are some known issues with Kotlin 2 and Skip-transpiled code right now, but we should be able to co-exist side-by-side until we work out the issues. I'll look into it today.

**Reply by anjumkaiser** [reply]
ok

I think docs should be updated to mention a way, people with a new setup might need help with this.

steps to reproduce:

macOS Sonoma 14.5
XCode 15.4

brew install openjdk -> install openjdk v22
brew install gradle -> installs grade 8.8
brew install kotlin -> installs kotlin v2.0

and the issue pops up with skip init

thank you

**Reply by marcprux** [reply]
I just tried upgrading to openjdk@22 / gradle@8.8 / kotlin 2, and Skip does still work, provided the openjdk@21 is still installed (which it should be, since we depending on `openjdk@21` on the Skip homebrew formula). I'll look into it some more, though – ideally, we would just work with whatever the latest openjdk is and not have to worry about it. But we did pin it to openjdk@21 when we migrated from openjdk@17 to ensure that the latest openjdk was going to be used.

**Reply by marcprux** [reply]
BTW, if you manually do `brew install openjdk@21`, does `skip checkup` pass for you?

**Reply by anjumkaiser** [reply]
I updated to latest skip, now checkup passes ok


skip checkup
[✓] Skip version 0.8.49 (= 0.8.49)
[✓] macOS version 14.5 (> 13.5.0)
[✓] Swift version 5.10 (> 5.9.0)
[✓] Xcode version 15.4 (> 15.0.0)
[✓] Xcode tools SDKs: 7
[✓] Homebrew version 4.3.4 (> 4.1.0)
[✓] Gradle version 8.8 (> 8.6.0)
[✓] Java version 22.0.1 (> 21.0.0)
[✓] Android Debug Bridge version 1.0.41 (> 1.0.40)
[✓] Android Studio version: 2023.3
[✓] Android tools SDKs: 7
[✓] Resolve dependencies (19.12s)
[✓] Build hello-skip (10.77s)
[✓] Test Swift (5.8s)
[✓] Test Kotlin (32.86s)
[✓] Archive iOS ipa (26.88s)
[✓] Assemble HelloSkip-release.ipa 37 KB
[✓] Verify HelloSkip-release.ipa 37 KB
[✓] Assembling Android apk (108.48s)
[✓] Verify HelloSkip-release.apk 12.7 MB
[✓] Check Swift Package (0.39s)
[✓] Skip 0.8.49 checkup succeeded in 204.93s

initialised a new project
\skip init --open-xcode --appid=com.xyz.HelloSkip hello-skip HelloSkip
Initializing Skip application hello-skip
[✓] Resolve dependencies (17.63s)
[✓] Build hello-skip (10.99s)
[✓] Check Swift Package (0.23s)
[✓] Created module HelloSkip in /Users/anjumkaiser/Developer/Skip/hello-skip
[✓] Opening Xcode project (0.25s)
[✓] Skip 0.8.49 init succeeded in 29.13s

it starts Xcode ok
I set team for signing and enabled skip plugin for Xcode

I started build from Xcode, it still fails

here is build export:
[Build HelloSkip_2024-06-11T04-42-46.txt](https://github.com/user-attachments/files/15780202/Build.HelloSkip_2024-06-11T04-42-46.txt)

**Reply by aabewhite** [reply]
The underlying error from the log is: 

GRADLE>    > Failed to calculate the value of task ':skipstone:SkipLib:compileDebugJavaWithJavac' property 'javaCompiler'.
GRADLE>       > Cannot find a Java installation on your machine matching this tasks requirements: {languageVersion=21, vendor=any, implementation=vendor-specific} for MAC_OS on aarch64.
GRADLE>          > No locally installed toolchains match and toolchain download repositories have not been configured.

**Comment by anjumkaiser** [comment]
I'm able to build in 0.8.51

**Reply by anjumkaiser** [reply]
thanks
